example/multistep/multistep.py

```diff
@@ -0,0 +1,92 @@
+"""
+Downloads the movielens dataset, ETLs it into Parquet, trains an
+ALS model, and uses the ALS model to train a Keras neural network.
+
+See README.rst for more details.
+"""
+
+import click
+import os
+import tempfile
+
+import mlflow
+from mlflow.entities import RunStatus, Run
+from mlflow.utils.logging_utils import eprint
+import six
+
+from mlflow.tracking.fluent import _get_experiment_id
+
+
+@click.group()
+def cli():
+    pass
+
+
+def _get_params(run):
+    """Converts [mlflow.entities.Param] to a dictionary of {k: v}."""
+    return {param.key: param.value for param in run.data.params}
+
+
+def _already_ran(entry_point_name, parameters, experiment_id=None):
+    """Best-effort detection of if a run with the given entrypoint name,
+    parameters, and experiment id already ran. The run must have completed
+    successfully and have at least the parameters provided.
+    """
+    experiment_id = experiment_id if experiment_id is not None else _get_experiment_id()
+    service = mlflow.tracking.get_service()
+    all_run_infos = reversed(service.list_run_infos(experiment_id))
+    for run_info in all_run_infos:
+        if run_info.entry_point_name != entry_point_name:
+            continue
+
+        full_run = service.get_run(run_info.run_uuid)
+        run_params = _get_params(full_run)
+        match_failed = False
+        for param_key, param_value in six.iteritems(parameters):
+            run_value = run_params.get(param_key)
+            if run_value != param_value:
+                match_failed = True
+                break
+        if match_failed:
+            continue
+
+        if run_info.status != RunStatus.FINISHED:
+            eprint(("Run matched, but is not FINISHED, so skipping "
+                    "(run_id=%s, status=%s)") % (run_info.run_uuid, run_info.status))
+            continue
+        return service.get_run(run_info.run_uuid)
+    return None
+
+
+# TODO(aaron): This is not great because it doesn't account for:
+# - changes in code
+# - changes in dependant steps
+def _get_or_run(entrypoint, parameters, use_cache=True):
+    existing_run = _already_ran(entrypoint, parameters)
+    if use_cache and existing_run:
+        print("Found existing run for entrypoint=%s and parameters=%s" % (entrypoint, parameters))
+        return existing_run
+    print("Launchng new run for entrypoint=%s and parameters=%s" % (entrypoint, parameters))
+    submitted_run = mlflow.run(".", entrypoint, parameters=parameters)
+    return mlflow.tracking.get_service().get_run(submitted_run.run_id)
+
+
+@cli.command()
+def workflow():
+    # Note: The entrypoint names are defined in MLproject. The artifact directories
+    # are documented by each step's .py file.
+    load_raw_data_run = _get_or_run("load_raw_data", {})
```

Comment:
This workflow is manually created based on hard coded steps. Would be awesome to be able to generate these DAGs/workflows directly from `MLproject` file. We could use airflow like syntax : `start`, `stop`, `upwards` or `downward` dependencies and generate the workflows from entry points. We might needs a few more hooks like input and output params to handle it.

Don't need to hold this PR. Probably another PR do such DAG building.

Reply:
Yeah, good point -- right now it's pretty prone to typos/changes. The goal here is to show what's possible today, and we should prioritize work to make it easier and less error-prone in the future.